crates/subspace-node/src/import_blocks_from_network.rs

```diff
@@ -0,0 +1,275 @@
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use clap::Parser;
+use log::{info, warn};
+use parity_scale_codec::Encode;
+use sc_cli::{CliConfiguration, ImportParams, SharedParams};
+use sc_client_api::{BlockBackend, HeaderBackend};
+use sc_consensus::{BlockImportError, BlockImportStatus, IncomingBlock, Link};
+use sc_service::ImportQueue;
+use sp_consensus::BlockOrigin;
+use sp_runtime::generic::BlockId;
+use sp_runtime::traits::{Block as BlockT, Header, NumberFor};
+use std::sync::Arc;
+use std::task::Poll;
+use subspace_archiving::reconstructor::Reconstructor;
+use subspace_core_primitives::Piece;
+use subspace_networking::libp2p::Multiaddr;
+use subspace_networking::{multimess, Config};
+use subspace_runtime_primitives::{RECORDED_HISTORY_SEGMENT_SIZE, RECORD_SIZE};
+
+type PieceIndex = u64;
+
+/// The `import-blocks-from-network` command used to import blocks from Subspace Network DSN.
+#[derive(Debug, Parser)]
+pub struct ImportBlocksFromNetworkCmd {
+    /// Multiaddrs of bootstrap nodes to connect to on startup, multiple are supported
+    #[clap(long)]
+    pub bootstrap_node: Vec<Multiaddr>,
+
+    /// The default number of 64KB pages to ever allocate for Wasm execution.
+    ///
+    /// Don't alter this unless you know what you're doing.
+    #[clap(long, value_name = "COUNT")]
+    pub default_heap_pages: Option<u32>,
+
+    #[allow(missing_docs)]
+    #[clap(flatten)]
+    pub shared_params: SharedParams,
+
+    #[allow(missing_docs)]
+    #[clap(flatten)]
+    pub import_params: ImportParams,
+}
+
+impl ImportBlocksFromNetworkCmd {
+    /// Run the import-blocks command
+    pub async fn run<B, C, IQ>(&self, client: Arc<C>, import_queue: IQ) -> sc_cli::Result<()>
+    where
+        C: HeaderBackend<B> + BlockBackend<B> + Send + Sync + 'static,
+        B: BlockT + for<'de> serde::Deserialize<'de>,
+        IQ: sc_service::ImportQueue<B> + 'static,
+    {
+        import_blocks(self.bootstrap_node.clone(), client, import_queue, false)
+            .await
+            .map_err(Into::into)
+    }
+}
+
+impl CliConfiguration for ImportBlocksFromNetworkCmd {
+    fn shared_params(&self) -> &SharedParams {
+        &self.shared_params
+    }
+
+    fn import_params(&self) -> Option<&ImportParams> {
+        Some(&self.import_params)
+    }
+}
+
+struct WaitLink {
+    imported_blocks: u64,
+    has_error: bool,
```

Comment:
Wrap the error directly? So that it can bubble up later.

```suggestion
    has_error: Option<BlockImportError>,
```

Reply:
IT was taken from Substrate, but I do agree that is better, refactrored